arelle/ValidateDuplicateFacts.py

```diff
@@ -167,6 +172,105 @@ def areWithinRoundingError(self) -> bool:
                 return False
         return True
 
+    def deduplicateCompleteSubsets(self) -> list[ModelFact]:
+        """
+        :return: A list of the first fact found for each unique decimals/value combination.
+        """
+        seenKeys = set()
+        results = []
+        for fact in self.facts:
+            key = (fact.decimals, fact.xValue)
+            if key in seenKeys:
+                continue
+            seenKeys.add(key)
+            results.append(fact)
+        return results
+
+    def deduplicateConsistentPairs(self) -> list[ModelFact]:
+        """
+        First performs deduplication of complete duplicates,
+        then removes from the remaining facts any fact that is consistent with a higher precision fact.
+        :return: A subset of the facts where the fact of lower precision in every consistent pair has been removed.
+        """
+        facts = self.deduplicateCompleteSubsets()
+        if not self.areNumeric:
+            # Consistency is equivalent to completeness for non-numeric facts
+            return facts
+        decimalsMap = defaultdict(list)
+        for fact in facts:
+            decimals = self.getDecimals(fact)
+            assert decimals is not None
+            decimalsMap[decimals].append(fact)
+        sortedDecimals = sorted(decimalsMap.keys())
+        results = set(facts)
+
+        for a, decimalsA in enumerate(sortedDecimals[:len(sortedDecimals)]):
```

Comment:
This implementation (`consistent-pairs` deduplication) delivers the end result of having no consistent duplicates in the output. I think "removing all consistent duplicate facts" is a reasonable interpretation of "consistent deduplication". Even though it's a higher time-complexity implementation, it seems fair to provide this option for anyone expecting this deduplication result.

The other implementation (`consistent-sets` deduplication) may retain consistent duplicates in the output since it takes a hands-off approach when inconsistent duplicates are encountered. We could rename the arg to `consistent` to nudge users in that direction if they're not sure.

Reply:
> This implementation (consistent-pairs deduplication) delivers the end result of having no consistent duplicates in the output

The only difference between `consistent-sets` and `consistent-pairs` is when the report has inconsistent duplicates, which means the report is already of "questionable" validity.  If a user asks to deduplicate, they probably consider inconsistent duplicates to be a fatal error, so I'm trying to understand in what scenario they would care about deduplicating amongst the inconsistent duplicates.  Giving an option that is equivalent to `consistent-sets` when the report is good but runs slower when the report needs more changes seems odd to me.

> We could rename the arg to consistent to nudge users in that direction if they're not sure.

Renaming `consistent-sets` to the shorter `consistent` seems good to me.  I'm not sure about `consistent-pairs` since a1,a2,a3 and b1,b2,b3 with values a,b at three varying precision levels still reduces to a3,b3, so at a high-level it's not about "pairs", it's that it tries hard to remove consistent duplicates even if there are inconsistent duplicates.  Perhaps `complete`, `consistent`, and `inconsistent` are reasonable names for what they do (particularly if `inconsistent` is documented as being potentially slow)?

Anyway, I'll let you all decide.